File: ToDoList/app/src/main/java/com/example/todolist/mactivity/AddTaskActivity.java
package com.example.todolist.mactivity;

import android.annotation.SuppressLint;
import android.app.TimePickerDialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RatingBar;
import android.widget.Spinner;
import android.widget.Toast;
import android.widget.ViewFlipper;

import com.example.todolist.R;
import com.example.todolist.madapter.SpinnerCategoryAdapter;
import com.example.todolist.mclass.Category;
import com.example.todolist.mclass.Task;
import com.example.todolist.mmodel.CategoryModel;
import com.example.todolist.mmodel.TaskModel;
import com.example.todolist.msqlite.SQLite;

import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Random;

public class AddTaskActivity extends AppCompatActivity {
    private ViewFlipper viewFlipperImage;
    private Toolbar myToolBar;
    private Spinner spinnerCategory;
    private TimePickerDialog timePickerDialog;
    private Button btnCreate;
    private EditText edtNameTask;
    private Calendar calendar;
    private RatingBar ratingTask;
    private String[] itemsIcon;
    private String type;
    private Category category;
    private String currentID;
    private SpinnerCategoryAdapter spinnerCategoryAdapter;
    private Task taskID;
    private SQLite sqLite;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_task);
        sqLite = new SQLite(getApplicationContext());

        findViewByIDs();
        addAppBar();

        spinnerCategoryAdapter = new SpinnerCategoryAdapter(AddTaskActivity.this, R.layout.spinner_category, CategoryModel.getItems());
        spinnerCategory.setAdapter(spinnerCategoryAdapter);
        addSpinner(false);

        addTime();
        addViewFlipperIcons(R.array.task_icon);

        currentID = getIntent().getStringExtra("getID");
        if (currentID != null) {
            addValue();
        }
    }

    @SuppressLint("SetTextI18n")
    private void addValue() {
        taskID = TaskModel.getItemsByID(currentID);

        assert taskID != null;
        edtNameTask.setText(taskID.getName());

        @SuppressLint("SimpleDateFormat") SimpleDateFormat simpleDateFormat = new SimpleDateFormat("HH:mm");
        btnCreate.setText("Starting at: " + simpleDateFormat.format(taskID.getTime().getTime()));

        ratingTask.setRating(taskID.getRating());

        int position = spinnerCategoryAdapter.getPosition(CategoryModel.getItemsByID(taskID.getType()));
        spinnerCategory.setSelection(position);

        addSpinner(true);
    }

    private void addSpinner(boolean flag) {
        spinnerCategory.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                category = (Category) spinnerCategory.getSelectedItem();
                type = category.getId();

                switch (type) {
                    case "Category_Home":
                        viewFlipperImage.removeAllViews();
                        addViewFlipperIcons(R.array.home_icon);
                        if (flag) {
                            String iconName = getResources().getResourceEntryName(taskID.getIcon());
                            int viewFlipperPos = Arrays.asList(itemsIcon).indexOf(iconName);
                            viewFlipperImage.setDisplayedChild(viewFlipperPos);
                        }
                        break;
                    case "Category_Working":
                        viewFlipperImage.removeAllViews();
                        addViewFlipperIcons(R.array.work_icon);
                        if (flag) {
                            String iconName = getResources().getResourceEntryName(taskID.getIcon());
                            int viewFlipperPos = Arrays.asList(itemsIcon).indexOf(iconName);
                            viewFlipperImage.setDisplayedChild(viewFlipperPos);
                        }
                        break;
                    case "Category_Sport":
                        viewFlipperImage.removeAllViews();
                        addViewFlipperIcons(R.array.sport_icon);
                        if (flag) {
                            String iconName = getResources().getResourceEntryName(taskID.getIcon());
                            int viewFlipperPos = Arrays.asList(itemsIcon).indexOf(iconName);
                            viewFlipperImage.setDisplayedChild(viewFlipperPos);
                        }
                        break;
                    case "Category_Entertainment":
                        viewFlipperImage.removeAllViews();
                        addViewFlipperIcons(R.array.entertainment_item);
                        if (flag) {
                            String iconName = getResources().getResourceEntryName(taskID.getIcon());
                            int viewFlipperPos = Arrays.asList(itemsIcon).indexOf(iconName);
                            viewFlipperImage.setDisplayedChild(viewFlipperPos);
                        }
                        break;
                    case "Category_Others":
                        viewFlipperImage.removeAllViews();
                        addViewFlipperIcons(R.array.other_item);
                        if (flag) {
                            String iconName = getResources().getResourceEntryName(taskID.getIcon());
                            int viewFlipperPos = Arrays.asList(itemsIcon).indexOf(iconName);
                            viewFlipperImage.setDisplayedChild(viewFlipperPos);
                        }
                        break;
                    case "Category_Studying":
                        viewFlipperImage.removeAllViews();
                        addViewFlipperIcons(R.array.study_item);
                        if (flag) {
                            String iconName = getResources().getResourceEntryName(taskID.getIcon());
                            int viewFlipperPos = Arrays.asList(itemsIcon).indexOf(iconName);
                            viewFlipperImage.setDisplayedChild(viewFlipperPos);
                        }
                        break;
                    default:
                        viewFlipperImage.removeAllViews();
                        addViewFlipperIcons(R.array.task_icon);
                        if (flag) {
                            String iconName = getResources().getResourceEntryName(taskID.getIcon());
                            int viewFlipperPos = Arrays.asList(itemsIcon).indexOf(iconName);
                            viewFlipperImage.setDisplayedChild(viewFlipperPos);
                        }
                        break;
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });
    }

    @SuppressLint("SetTextI18n")
    private void findViewByIDs() {
        viewFlipperImage = findViewById(R.id.viewFlipperImage);
        myToolBar = findViewById(R.id.myAddToolBar);
        spinnerCategory = findViewById(R.id.spinnerCategory);

        btnCreate = findViewById(R.id.btnCreate);
        calendar = Calendar.getInstance();
        @SuppressLint("SimpleDateFormat") SimpleDateFormat simpleDateFormat = new SimpleDateFormat("HH:mm");
        btnCreate.setText("Starting at: " + simpleDateFormat.format(calendar.getTime()));

        edtNameTask = findViewById(R.id.edtNameTask);
        ratingTask = findViewById(R.id.ratingTask);
    }

    private void addAppBar() {
        setSupportActionBar(myToolBar);
        ActionBar actionBar = this.getSupportActionBar();
        assert actionBar != null;
        actionBar.setDisplayHomeAsUpEnabled(true); //thêm nút back
    }

    private void addViewFlipperIcons(int allItems) {
        itemsIcon = getResources().getStringArray(allItems);
        for (String icons : itemsIcon) {
            viewFlipperImage.addView(new ViewFlipperCustomActivity(AddTaskActivity.this, icons));
        }
    }

    @SuppressLint("SetTextI18n")
    private void addTime() {
        calendar = Calendar.getInstance();
        timePickerDialog = new TimePickerDialog(AddTaskActivity.this, (view, hourOfDay, minute) -> {
            calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
            calendar.set(Calendar.MINUTE, minute);

            @SuppressLint("SimpleDateFormat") SimpleDateFormat simpleDateFormat = new SimpleDateFormat("HH:mm");
            btnCreate.setText("Starting at: " + simpleDateFormat.format(calendar.getTime()));
        }, calendar.get(Calendar.HOUR), calendar.get(Calendar.MINUTE), true);
    }

    @SuppressLint("NonConstantResourceId")
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.imgBtnPrevious:
                viewFlipperImage.showPrevious();
                break;
            case R.id.imgBtnNext:
                viewFlipperImage.showNext();
                break;
            case R.id.btnCreate:
                timePickerDialog.show();
                break;
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_add, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.action_check) {

            String name = edtNameTask.getText().toString();
            if (TextUtils.isEmpty(name)) {
                Toast.makeText(AddTaskActivity.this, "Please enter your task name!!!", Toast.LENGTH_LONG).show();
            } else {
                String ID = "Task_" + randomString();
                int rating = (int) ratingTask.getRating();
                int icon = ViewFlipperCustomActivity.getResource(AddTaskActivity.this, itemsIcon[viewFlipperImage.getDisplayedChild()]);
                if (currentID != null) {
                    TaskModel.editTask(currentID, name, calendar, icon, rating, type);
                    Task currentTask = TaskModel.getItemsByID(currentID);
                    sqLite.edit(currentTask);
                    setResult(MainActivity.REQUEST_CODE_TASK_EDIT_DONE, getIntent());
                } else {
                    Task task = new Task(ID, name, calendar, icon, rating, type);
                    if (sqLite.add(task)) {
                        Toast.makeText(AddTaskActivity.this, "Success!!!", Toast.LENGTH_SHORT).show();
                    } else {
                        Toast.makeText(AddTaskActivity.this, "Fail!!!", Toast.LENGTH_SHORT).show();
                    }
                    TaskModel.addTask(ID, name, calendar, icon, rating, type);
                    setResult(MainActivity.REQUEST_CODE_TASK_ADD_DONE, getIntent());
                }
                //Toast.makeText(AddTaskActivity.this, "Successful!!!", Toast.LENGTH_LONG).show();
            }
        }
        return super.onOptionsItemSelected(item);
    }

    private String randomString() {
        String characters = "abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        int characters_length = characters.length();
        Random random = new Random();
        StringBuilder id = new StringBuilder();
        for (int i = 0; i < 4; i++) {
            id.append(characters.charAt(random.nextInt(characters_length)));
        }
        return id.toString();
    }
}

File: ToDoListKotlin/app/src/main/java/com/example/todolistkotlin/activity/activity/AddTaskActivity.kt
package com.example.todolistkotlin.activity.activity

import android.annotation.SuppressLint
import android.app.TimePickerDialog
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import android.widget.AdapterView.OnItemSelectedListener
import com.example.todolistkotlin.activity.adapter.SpinnerCategoryAdapter
import com.example.todolistkotlin.activity.classes.Category
import com.example.todolistkotlin.activity.classes.Task
import com.example.todolistkotlin.activity.model.CategoryModel
import com.example.todolistkotlin.activity.model.TaskModel
import com.example.todolistkotlin.activity.sqlite.SQLite
import com.example.todolistkotlin.R
import kotlinx.android.synthetic.main.activity_add_task.*
import java.text.SimpleDateFormat
import java.util.*


class AddTaskActivity : AppCompatActivity() {
    private var timePickerDialog: TimePickerDialog? = null
    private var calendar: Calendar? = null
    private lateinit var itemsIcon: Array<String>
    private var type: String? = null
    private var category: Category? = null
    private var currentID: String? = null
    private var spinnerCategoryAdapter: SpinnerCategoryAdapter? = null
    private var taskID: Task? = null
    private var sqLite: SQLite? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_task)
        sqLite = SQLite(applicationContext)
        findViewByIDs()
        addAppBar()
        spinnerCategoryAdapter = CategoryModel.getItems()?.let {
            SpinnerCategoryAdapter(
                    this@AddTaskActivity,
                    R.layout.spinner_category,
                    it
        )
        }
        spinnerCategory!!.adapter = spinnerCategoryAdapter
        addSpinner(false)
        addTime()
        addViewFlipperIcons(R.array.task_icon)
        currentID = intent.getStringExtra("getID")
        if (currentID != null) {
            addValue()
        }
    }

    @SuppressLint("SetTextI18n")
    private fun addValue() {
        taskID = TaskModel.getItemsByID(currentID)
        edtNameTask.setText(taskID!!.getName())
        @SuppressLint("SimpleDateFormat") val simpleDateFormat = SimpleDateFormat("HH:mm")
        btnCreate!!.text = "Starting at: " + simpleDateFormat.format(taskID!!.getTime().time)
        ratingTask!!.rating = taskID!!.getRating().toFloat()
        val position: Int? = spinnerCategoryAdapter?.getPosition(CategoryModel.getItemsByID(taskID!!.getType()))
        if (position != null) {
            spinnerCategory!!.setSelection(position)
        }
        addSpinner(true)
    }

    private fun addSpinner(flag: Boolean) {
        spinnerCategory!!.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View,
                position: Int,
                id: Long
            ) {
                category = spinnerCategory!!.selectedItem as Category
                type = category!!.getId()
                when (type) {
                    "Category_Home" -> {
                        viewFlipperImage!!.removeAllViews()
                        addViewFlipperIcons(R.array.home_icon)
                        if (flag) {
                            val iconName = taskID?.let { resources.getResourceEntryName(it.getIcon()) }
                            val viewFlipperPos = listOf(*itemsIcon).indexOf(iconName)
                            viewFlipperImage!!.displayedChild = viewFlipperPos
                        }
                    }
                    "Category_Working" -> {
                        viewFlipperImage!!.removeAllViews()
                        addViewFlipperIcons(R.array.work_icon)
                        if (flag) {
                            val iconName = taskID?.let { resources.getResourceEntryName(it.getIcon()) }
                            val viewFlipperPos = listOf(*itemsIcon).indexOf(iconName)
                            viewFlipperImage!!.displayedChild = viewFlipperPos
                        }
                    }
                    "Category_Sport" -> {
                        viewFlipperImage!!.removeAllViews()
                        addViewFlipperIcons(R.array.sport_icon)
                        if (flag) {
                            val iconName = taskID?.let { resources.getResourceEntryName(it.getIcon()) }
                            val viewFlipperPos = listOf(*itemsIcon).indexOf(iconName)
                            viewFlipperImage!!.displayedChild = viewFlipperPos
                        }
                    }
                    "Category_Entertainment" -> {
                        viewFlipperImage!!.removeAllViews()
                        addViewFlipperIcons(R.array.entertainment_item)
                        if (flag) {
                            val iconName = taskID?.let { resources.getResourceEntryName(it.getIcon()) }
                            val viewFlipperPos = listOf(*itemsIcon).indexOf(iconName)
                            viewFlipperImage!!.displayedChild = viewFlipperPos
                        }
                    }
                    "Category_Others" -> {
                        viewFlipperImage!!.removeAllViews()
                        addViewFlipperIcons(R.array.other_item)
                        if (flag) {
                            val iconName = taskID?.let { resources.getResourceEntryName(it.getIcon()) }
                            val viewFlipperPos = listOf(*itemsIcon).indexOf(iconName)
                            viewFlipperImage!!.displayedChild = viewFlipperPos
                        }
                    }
                    "Category_Studying" -> {
                        viewFlipperImage!!.removeAllViews()
                        addViewFlipperIcons(R.array.study_item)
                        if (flag) {
                            val iconName = taskID?.let { resources.getResourceEntryName(it.getIcon()) }
                            val viewFlipperPos = listOf(*itemsIcon).indexOf(iconName)
                            viewFlipperImage!!.displayedChild = viewFlipperPos
                        }
                    }
                    else -> {
                        viewFlipperImage!!.removeAllViews()
                        addViewFlipperIcons(R.array.task_icon)
                        if (flag) {
                            val iconName = taskID?.let { resources.getResourceEntryName(it.getIcon()) }
                            val viewFlipperPos = listOf(*itemsIcon).indexOf(iconName)
                            viewFlipperImage!!.displayedChild = viewFlipperPos
                        }
                    }
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    @SuppressLint("SetTextI18n")
    private fun findViewByIDs() {
        calendar = Calendar.getInstance()
        @SuppressLint("SimpleDateFormat") val simpleDateFormat = SimpleDateFormat("HH:mm")
        btnCreate.text = "Starting at: " + simpleDateFormat.format(calendar!!.time)
    }

    private fun addAppBar() {
        setSupportActionBar(myAddToolBar)
        val actionBar = this.supportActionBar!!
        actionBar.setDisplayHomeAsUpEnabled(true) //thêm nút back
    }

    private fun addViewFlipperIcons(allItems: Int) {
        itemsIcon = resources.getStringArray(allItems)
        for (icons in itemsIcon) {
            viewFlipperImage!!.addView(ViewFlipperCustomActivity(this@AddTaskActivity, icons))
        }
    }

    @SuppressLint("SetTextI18n")
    private fun addTime() {
        calendar = Calendar.getInstance()
        timePickerDialog = TimePickerDialog(
            this@AddTaskActivity,
            { _: TimePicker?, hourOfDay: Int, minute: Int ->
                calendar!!.set(Calendar.HOUR_OF_DAY, hourOfDay)
                calendar!!.set(Calendar.MINUTE, minute)
                @SuppressLint("SimpleDateFormat") val simpleDateFormat =
                    SimpleDateFormat("HH:mm")
                btnCreate!!.text = "Starting at: " + simpleDateFormat.format(calendar!!.time)
            }, calendar!!.get(Calendar.HOUR), calendar!!.get(Calendar.MINUTE), true
        )
    }

    @SuppressLint("NonConstantResourceId")
    fun onClick(view: View) {
        when (view.id) {
            R.id.imgBtnPrevious -> viewFlipperImage!!.showPrevious()
            R.id.imgBtnNext -> viewFlipperImage!!.showNext()
            R.id.btnCreate -> timePickerDialog!!.show()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_add, menu)
        return super.onCreateOptionsMenu(menu)
    }

    @SuppressLint("NonConstantResourceId")
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_check) {
            val name = edtNameTask!!.text.toString()
            if (TextUtils.isEmpty(name)) {
                Toast.makeText(
                    this@AddTaskActivity,
                    "Please enter your task name!!!",
                    Toast.LENGTH_LONG
                ).show()
            } else {
                val ID = "Task_" + randomString()
                val rating = ratingTask!!.rating.toInt()
                val icon: Int = ViewFlipperCustomActivity.getResource(
                    this@AddTaskActivity,
                    itemsIcon[viewFlipperImage!!.displayedChild]
                )
                if (currentID != null) {
                    TaskModel.editTask(currentID, name, calendar, icon, rating, type)
                    val currentTask: Task? = TaskModel.getItemsByID(currentID)
                    if (currentTask != null) {
                        sqLite?.edit(currentTask)
                    }
                    setResult(MainActivity.REQUEST_CODE_TASK_EDIT_DONE, intent)
                } else {
                    val task = Task(ID, name, calendar!!, icon, rating, type!!)
                    if (sqLite?.add(task) == true) {
                        Toast.makeText(this@AddTaskActivity, "Success!!!", Toast.LENGTH_SHORT)
                            .show()
                    } else {
                        Toast.makeText(this@AddTaskActivity, "Fail!!!", Toast.LENGTH_SHORT).show()
                    }
                    TaskModel.addTask(ID, name, calendar, icon, rating, type)
                    setResult(MainActivity.REQUEST_CODE_TASK_ADD_DONE, intent)
                }
                //Toast.makeText(AddTaskActivity.this, "Successful!!!", Toast.LENGTH_LONG).show();
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun randomString(): String {
        val characters = "abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        val characters_length = characters.length
        val random = Random()
        val id = StringBuilder()
        for (i in 0..3) {
            id.append(characters[random.nextInt(characters_length)])
        }
        return id.toString()
    }
}